src/lib/routes/admin-api/public-signup.test.ts

```diff
@@ -0,0 +1,211 @@
+        const appName = '123!23';
+
+        stores.clientApplicationsStore.upsert({ appName });
+        stores.publicSignupTokenStore.create({
+            name: 'some-name',
+            expiresAt: expireAt(),
+        });
+
+        return request
+            .get('/api/admin/invite-link/tokens/some-secret')
+            .expect(200)
+            .expect((res) => {
+                const token = res.body;
+                expect(token.name).toBe('some-name');
+            });
+    });
+
+    test('should expire token', async () => {
+        expect.assertions(1);
+        const appName = '123!23';
+
+        stores.clientApplicationsStore.upsert({ appName });
+        stores.publicSignupTokenStore.create({
+            name: 'some-name',
+            expiresAt: expireAt(),
+        });
+
+        return request
+            .delete('/api/admin/invite-link/tokens/some-secret')
+            .expect(200)
+            .expect(async () => {
+                const eventCount = await stores.eventStore.count();
+                expect(eventCount).toBe(1); // PUBLIC_SIGNUP_TOKEN_MANUALLY_EXPIRED
+            });
+    });
+
+    test('should create user and add to token', async () => {
+        expect.assertions(3);
+        const appName = '123!23';
+
+        stores.clientApplicationsStore.upsert({ appName });
+        stores.publicSignupTokenStore.create({
+            name: 'some-name',
+            expiresAt: expireAt(),
+        });
+
+        const user: CreateUserSchema = {
+            username: 'some-username',
+            email: 'someEmail@example.com',
+            name: 'some-name',
+            password: null,
+            rootRole: 1,
+            sendEmail: false,
+        };
+
+        return request
+            .post('/api/admin/invite-link/tokens/some-secret/signup')
+            .send(user)
+            .expect(201)
+            .expect(async (res) => {
+                const count = await stores.userStore.count();
+                expect(count).toBe(1);
+                const eventCount = await stores.eventStore.count();
+                expect(eventCount).toBe(2); //USER_CREATED && PUBLIC_SIGNUP_TOKEN_USER_ADDED
+                expect(res.body.username).toBe(user.username);
+            });
+    });
+
+    test('should return 200 if token is valid', async () => {
+        const appName = '123!23';
+
+        stores.clientApplicationsStore.upsert({ appName });
+        stores.publicSignupTokenStore.create({
+            name: 'some-name',
+            expiresAt: expireAt(),
+        });
+
+        return request
+            .post('/api/admin/invite-link/tokens/some-secret/validate')
+            .expect(200);
+    });
+
+    test('should return 404 if token is invalid', async () => {
+        const appName = '123!23';
+
+        stores.clientApplicationsStore.upsert({ appName });
+
+        return request
+            .post('/api/admin/invite-link/tokens/some-invalid-secret/validate')
+            .expect(404);
```

Comment:
This should be 401. 404 should only be used if the server can't find a resource matching the URL: https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html

Reply:
fixed